R/teal_slice.R

```diff
@@ -0,0 +1,431 @@
+#' `teal_slices` also specifies which variables cannot be filtered
+#' and how observations are tallied, which is resolved by `FilterStates`.
+#'
+#' @param dataname `character(1)` name of data set
+#' @param varname `character(1)` name of variable
+#' @param choices vector specifying allowed choices;
+#'                possibly a subset of values in data;
+#'                type and size depends on variable type
+#' @param selected vector specifying selection;
+#'                 type and size depends on variable type
+#' @param keep_na `logical(0-1)` optional logical flag specifying whether to keep missing values
+#' @param keep_inf `logical(0-1)` optional logical flag specifying whether to keep infinite values
+#' @param fixed `logical(1)` logical flag specifying whether to fix this filter state (i.e. forbid setting state)
+#' @param exclude `named list` of `character` vectors where list names match names of data sets
+#'                 and vector elements match variable names in respective data sets;
+#'                 specifies which variables are not allowed to be filtered
+#' @param count_type `character(1)` string specifying how observations are tallied by these filter states
+#' @param show_all `logical(1)` specifying whether NULL elements should also be printed
+#' @param tss `teal_slices`
+#' @param feature `character(1)` name of `teal_slice` element
+#' @param expr `expression` or `character` string representing and expression
+#'             that evaluates to a single `logical`; will be evaluated in individual `teal_slice` objects
+#' @param ... for `filter_var` any number of additional features given as `name:value` pairs\cr
+#'            for `filter_settings` any number of `teal_slice` objects\cr
+#'            for other functions arguments passed to other methods
+#'
+#' @return
+#' `filter_var` returns object of class `teal_slice`, which is a named list.
+#' `filter_settings` returns object of class `teal_slices`, which is an unnamed list of `teal_slice` objects.
+#'
+#' @examples
+#' filter_1 <- filter_var("dataname1", "varname1", letters, "b", FALSE, extra1 = "extraone")
+#' filter_2 <- filter_var("dataname1", "varname2", 1:10, 2, TRUE, FALSE, extra2 = "extratwo")
+#' filter_3 <- filter_var("dataname2", "varname3", 1:10/10, 0.2, TRUE, FALSE,
+#'                        extra1 = "extraone", extra2 = "extratwo")
+#'
+#' all_filters <- filter_settings(
+#'   filter_1,
+#'   filter_2,
+#'   filter_3,
+#'   exclude = list(
+#'     "dataname1" = "varname2"
+#'   )
+#' )
+#'
+#' teal.slice:::extract_by_feat(all_filters, "dataname", "dataname1")
+#' teal.slice:::extract_feat(all_filters, "dataname")
+#' teal.slice:::extract_fun(all_filters, dataname == "dataname2")
+#' teal.slice:::extract_fun_s(all_filters, 'dataname == "dataname2"')
+#' x <- "dataname2"
+#' teal.slice:::extract_fun_s(all_filters, sprintf('dataname == "%s"', x))
+#'
+#' @name teal_slice
+NULL
+
+
+#' @export
+#' @rdname teal_slice
+#'
+filter_var <- function(
```

Comment:
I get needing 'dataname' and 'varname' from the outside perspective (initializing an app). But it feels extra from an inside perspective (e.g. in FilterStates).

What if `filter_settings` took named `teal_slice`s? Kind of like now with the lists but the right hand side is a `filter_var` instead of a list. This might need another function though to handle all filters for a single data set.

I'm thinking something more like this.
```r
default_filters <- filter_settings(
    ADSL = filter_data( # filter_group maybe?
        filter_var(AGE, selected = c(0, 100)),
        filter_var(RACE)
    ),
    ADTTE = filter_data(
        filter_var(AVAL, keep_na = FALSE)
    )
)
```

Once FilteredData, FilterStates are initialized, an individual `filter_var` shouldn't care what the dataname or varname is. That part can be handled by the class via e.g. a named list where names correspond to the variable.

Otherwise if I want to set a single FilterState say AGE in ADSL, I have to call `fs$set_state(ADSL, AGE, selected = c(0, 10))`? The filter state `fs` knows what data and variable it belongs to though so it shouldn't have to specified here. I could be misunderstanding how this works though

Reply:
I disagree. 

1. To my understanding the goal of the new api is to pass complete information on a single `FilterState` through all classes of the filter panel.
2. Builtding a `teal_slice` in layers would be no different from the nested lists this is meant to replace. Why stop at `dataname`? There are other elements one could be adding along the way, e.g. what is currently called `varlabel` or `target`. (Note: `exclude` and `count_type` are departures from this, as is `extract_type`).
3. Nobody is going to type these specifications other than for the sake of examples or during app development.
4. API-driven calls to `FilterState` objects are called from the top level anyway.

In the context of (4), the `FilterState`s don't care which dataset they belong to, but in the unlikely event that the same variable (`varname`) name exists in more than one dataset (`dataname`), both are needed.

Ultimately, there are no solutions, only trade-offs. I should probably frame this on a wall somewhere as keeps coming up lately. To me having to type more is a reasonable trade-off for a `teal_slice` being an atom of information. (I suppose `dataname` and `varname` sould be the nucleus.) 

Furthermore, a `FilterState` not knowing or caring what `FilterStates` it belongs to keeps it simple, which we are in dire need of.

Reply:
@asbates @chlebowa Let's divide discussion on three parts which have been put on question (all can be independent). Point of discussion discussion is that all understand each other and understand the technical consequences of doing something in another way. If during discussion you figure out that it's okey let's merge it and continue discussion in another issue.

1. Public functions to interact with filter-panel. @asbates put recent implementation in question as `filter_data` might be a good addition.

2. Apart from (1), `teal_slices` might be divided per dataset list to avoid holding and specifying `dataname` information multiple times. This can be independent of (1) as returned `teal_slices` might be built in the same way as implemented already.

3. `FIlterState` isn't strictly tied up with `teal_slice` as initialize uses arguments specified separately and `set_state` requires `teal_slice`. Both can be viewed as inconsistent.

------

I'd like to hear others opinion about `initialize` and `set/get_state` in `FilterState` (3). Should `FilterState` return `teal_slice` or a `list` which is then turned into `teal_slice` in `FilterState`. How it's implemented on this branch is up for discussion:
- Is called on resolved `teal_slice` and `FilterState$initialize` receives all attributes in the arguments (not via teal_slice).
- `FIlterState$initialize` sets fixed features (dataname, varname, choices, fixed, extract_type, datalabel, target) and reactive features (selected, keep_na, keep_inf)
- `FilterState$set_state` accept `teal_slice` object and sets reactive features and ignores static features.
- `FIlterState$get_state` returns `teal_slice` as it holds all necessary informations. 

Alternatively:
- we can remove all `dataset` specific arguments from `init_filter_state` so that it will initialize with `varname, varlabel, choices, selected, keep_na, keep_inf`
- we can change `FilterState$set_state(state)` to accept a state=list containing `selected, keep_na, keep_inf`
- `FilterState$get_state` returns a list equivalent to `state` in `set_state`
- `FIlterStates$get_filter_state` would create `teal_slice` based on the list returned from `get_state` and on the dataset attributes (dataname, target).


-----------

My personal view - I'm fine with current implementation. There is one important reason (for me) why `FilterState$get_state` returns more informations (also irrelevant for itself). We can try to remove some of the unnecessary steps - for example `filter_state_list` holding two lists for summarized experiments.

Reply:
> I'd like to hear others opinion about `initialize` and `set/get_state` in `FilterState` (3). Should `FilterState` return `teal_slice` or a `list` which is then turned into `teal_slice` in `FilterState`. How it's implemented on this branch is up for discussion:
> 
> * Is called on resolved `teal_slice` and `FilterState$initialize` receives all attributes in the arguments (not via teal_slice).
> * `FIlterState$initialize` sets fixed features (dataname, varname, choices, fixed, extract_type, datalabel, target) and reactive features (selected, keep_na, keep_inf)

I don't understand what you mean by this.

Full disclosure: I am biased against reversing changes I myself made in this PR.

I don't really see what is gained by dropping `dataname` from a `teal_slice` other than a little convenience. On the other hand, here are the gains I see by not dropping it:
- unambiguous clarity of `FilterState` definition
- consistency of defining state across the hierarchy
- ease of restructure the hierarchy (which has not been mentioned in this PR specifically, but was implied in others)

Reply:
![local-filter-panel-api comment drawio](https://user-images.githubusercontent.com/6959016/226738835-8b6e2daf-d5c0-40aa-a393-307dffeb6f5f.png)


Sorry for a stream of consciousness. Problematic is what I drawed above. We initialize FS by explicitly passing arguments and set_choices, set_selected directly in init. `teal_slice` object is unwrapped in `FilterStates`. While `set_state` uses `teal_slice` directly. I think we can optimize this, but it doesn't mean we need to do this in this PR. Here are some propositions:

#### 1 Initialize creates "empty" FS 

Everything is set through set_state as a single point of truth. `teal_slice` is utilized and reproduced in the FilterState


![local-filter-panel-api 2 drawio (1)](https://user-images.githubusercontent.com/6959016/226740154-c3a543ed-73c7-418e-9fde-7ed1a600f5b8.png)




#### 2 compose/decompose teal_slice in FilterStates

`teal_slice` is utilized and reproduced in `FilterStates`. Thanks to this, `FilterState` doesn't store/return anything unnecessary 

![local-filter-panel-api 3 drawio](https://user-images.githubusercontent.com/6959016/226740395-78121520-f5bf-41c2-934e-1a8f8d2d34d2.png)


Let's discuss this tomorrow